### PR TITLE
Update hapi core modules

### PR DESCRIPTION
**Pull request by Samueljoli**
## Description
This pull request updates the hapi core modules joi and joi-date to the latest and their namespaced versions. Updating to lab v18.x.x [fixed](https://github.com/hapijs/lab/pull/878) an [issue](https://github.com/hapijs/lab/pull/878) with false positive code coverage of if statements. It seems like they've updated their linting rules as well. I added tests to bring us to 100% coverage and fixed lint errors along with this work.

## Types of changes

- [ ] Documentation only (no changes to either `lib/` or `test/` files)
- [ ] Bug fix (non-breaking change which fixes an issue. you didn't modify existing tests)
- [x] New feature (non-breaking change which adds functionality. you added at least one new test)
- [ ] Breaking change (fix or feature that would cause existing functionality to change. you had to modify existing tests.)

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/xogroup/felicity/blob/master/.github/CONTRIBUTING.md) document.
- [x] My code follows the [Hapi.js style guide](https://github.com/hapijs/contrib/blob/master/Style.md).
- [x] I have updated the documentation as needed.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: lib/helpers.js [hunk 1]
```diff
@@ -42,7 +42,7 @@ const descriptionCompiler = function (description) {
             if (flag === 'presence') {
                 base = base[description.flags[flag]]();
             }
-            else if (flag === 'allowOnly' && description.valids) {
```
Comment (Samueljoli) [inline lib/helpers.js]: Removed this additional check because there will never be a case where `.valid()` is used and both `allowOnly` and the `valids` array isn't truthy

File: lib/valueGenerator.js [hunk 2]
```diff
@@ -53,11 +66,13 @@ class Any {
...
             }
         }
+
         if (Hoek.reach(this, '_options.ignoreDefaults') !== true && Hoek.reach(schemaDescription, 'flags.default') !== undefined) {
             return this._getDefaults();
         }
+
         if (Hoek.reach(schemaDescription, 'examples')) {
-            return Helpers.pickRandomFromArray(schemaDescription.examples);
+            return Helpers.pickRandomFromArray(schemaDescription.examples).value;
```
Comment (Samueljoli) [inline lib/valueGenerator.js]: `any.example()` now returns an [object](https://github.com/hapijs/joi/blob/ea9facf4d2133c7667071328759bc84f42455f59/lib/types/any/index.js#L512):
```js
{ value: '53', options: {} } 
```

File: test/value_generator_tests.js [hunk 3]
```diff
@@ -1419,6 +1502,14 @@ describe('Object', () => {
         ExpectValidation(example, schema);
     });
 
+    it('should return an ojbect with at least a key or peer', () => {
```
Comment (Samueljoli) [inline test/value_generator_tests.js]: This test is kinda bogus. I don't think there would ever be logical case where someone would create a schema, use `with()` and not define both the key and peer in the schema [`lib/valueGenerator`](https://github.com/xogroup/felicity/blob/master/lib/valueGenerator.js#L728). More or less wrote this test for coverage. Open to thoughts. Got half a mind of removing the check.

File: package.json [hunk 4]
```diff
@@ -24,9 +24,9 @@
   },
   "homepage": "https://github.com/xogroup/felicity#readme",
   "dependencies": {
+    "@hapi/joi": "15.0.3",
```
Comment (WesTyler) [inline package.json]: these can be pinned just to the major version still
```
"@hapi/joi": "15.x",
"@hapi/joi-date": "1.x",
```

File: lib/valueGenerator.js (hunk 2 above)
Comment (WesTyler) [reply lib/valueGenerator.js]: Yeah that was a breaking change in Joi 14 that I never got around to updating to 😆

File: package.json [hunk 5]
```diff
@@ -35,8 +35,8 @@
     "joi": "13.x"
   },
   "devDependencies": {
-    "code": "5.x.x",
-    "lab": "15.x.x",
+    "@hapi/code": "5.3.1",
+    "@hapi/lab": "18.1.2",
```
Comment (WesTyler) [inline package.json]: Similarly pin to majors only

File: test/value_generator_tests.js [hunk 6]
```diff
@@ -1,14 +1,14 @@
...
 const Joi = require('../lib/joi');
-const Lab = require('lab');
-const Permutations = require('./test_helpers').permutations;
+const { permutations, expectValidation } = require('./test_helpers');
 const ValueGenerator = require('../lib/exampleGenerator');
-const Moment = require('moment');
 
 const { describe, expect, it } = exports.lab = Lab.script();
-const ExpectValidation = require('./test_helpers').expectValidation.bind({}, expect);
+const ExpectValidation = expectValidation(expect);
```
Comment (WesTyler) [inline test/value_generator_tests.js]: Good cleanup 👍

File: test/value_generator_tests.js [hunk 7]
```diff
@@ -1354,10 +1423,10 @@ describe('Object', () => {
 
     it('should return an object with min number of keys', () => {
 
-        const schema = Joi.object().min(5);
+        const schema = Joi.object().keys({ prop: Joi.string() }).min(1);
```
Comment (WesTyler) [inline test/value_generator_tests.js]: Was there a breakage that you noticed that you needed to add the `.keys({})` call? This test was specifically for the edge case of dynamic keys where the schema doesn't know what they are called but does know how many should be present.

File: test/value_generator_tests.js (hunk 3 above)
Comment (WesTyler) [reply test/value_generator_tests.js]: Nah I actually think it's a good test. Felicity supports several of these cases where it's "weird but technically valid".

File: lib/valueGenerator.js [hunk 8]
```diff
@@ -7,25 +7,36 @@ const Uuid = require('uuid');
 const Moment = require('moment');
 const Helpers = require('./helpers');
 
-const internals = {};
+const internals = {
+    JoiArrayProto : Reflect.getPrototypeOf(Joi.array())
```
Comment (WesTyler) [inline lib/valueGenerator.js]: missing a comma at the end that's preventing tests from running

**Comment by Samueljoli** [comment]
@WesTyler Ello, it's been a few, so just checking up on the status of this 😄
